vector/src/test/java/im/vector/app/features/crypto/UISIDetectorTest.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2022 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.features.crypto
+
+import im.vector.app.E2EMessageDetected
+import im.vector.app.UISIDetector
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.amshove.kluent.fail
+import org.junit.Assert
+import org.junit.Test
+import org.matrix.android.sdk.api.crypto.MXCRYPTO_ALGORITHM_MEGOLM
+import org.matrix.android.sdk.api.session.crypto.MXCryptoError
+import org.matrix.android.sdk.api.session.events.model.Event
+import org.matrix.android.sdk.api.session.events.model.EventType
+import org.matrix.android.sdk.api.session.events.model.content.EncryptedEventContent
+import org.matrix.android.sdk.api.session.events.model.toContent
+
+class UISIDetectorTest {
+
+    @Test
+    fun `trigger detection after grace period`() {
+        val uisiDetector = UISIDetector()
+        var detectedEvent: E2EMessageDetected? = null
+
+        uisiDetector.callback = object : UISIDetector.UISIDetectorCallback {
+            override val enabled = true
+            override val reciprocateToDeviceEventType = "foo"
+
+            override fun uisiDetected(source: E2EMessageDetected) {
+                detectedEvent = source
+            }
+
+            override fun uisiReciprocateRequest(source: Event) {
+                // nop
+            }
+        }
+
+        // report a decryption error
+        val eventId = "0001"
+        val event = fakeEncryptedEvent(eventId, "s1", "r1")
+        uisiDetector.onEventDecryptionError(event, fakeCryptoError())
+
+        runBlocking {
+            delay(40_000)
```

Comment:
The test will last at least 40 seconds? I think there is a way for test case like this much faster with coroutines (sort of fake time / accelerated time)

Reply:
The grace period is 30s. I could make it configurable and then reduce the test duration, but is it really an issue if it's taking 40s?

Reply:
we want test (and so the CI) to be fast!

Reply:
I made the grace period configurable, and reduced the test duration.